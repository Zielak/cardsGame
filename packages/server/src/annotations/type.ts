import {
  Context,
  DefinitionType,
  Schema,
  type as colType,
} from "@colyseus/schema"
// import { globalContext } from "@colyseus/schema/lib/annotations"

import { globalEntitiesContext } from "./entitiesContext.js"

export function type(
  typeDef: DefinitionType,
  typesContext?: Context
): PropertyDecorator {
  return function (target: typeof Schema, field: string): void {
    const constructor = target.constructor as typeof Schema
    const context = globalEntitiesContext

    // Intercept type definition
    if (field.indexOf("children") !== 0) {
      if (!context.registeredTypeDefinitions.has(constructor.name)) {
        context.registeredTypeDefinitions.set(constructor.name, new Map())
      }
      // logs.debug("child type:", `${constructor.name}.${field} ${logType}`)
      context.registeredTypeDefinitions
        .get(constructor.name)
        .set(field, typeDef)
    }

    return colType(typeDef, typesContext)(target, field)
  }
}

export function defineTypes(
  target: typeof Schema,
  fields: { [property: string]: DefinitionType },
  context?: Context // = globalContext
): typeof Schema {
  for (const field in fields) {
    type(fields[field], context)(target.prototype, field)
  }
  return target
}
