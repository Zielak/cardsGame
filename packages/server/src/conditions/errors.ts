import { getFlag } from "./utils.js"

export type CustomConditionError = string

/**
 * Use to fail an assertion other part of Condition for any reason (like validation).
 * Use custom error message if available
 * @category Conditions
 */
export function throwError(target: Record<string, any>, message: string): void {
  const customMessage = getCustomError(target)

  throw new Error(customMessage ?? message)
}

export function getCustomError(
  target: Record<string, any>
): CustomConditionError {
  return getFlag(target, "customError")
}
