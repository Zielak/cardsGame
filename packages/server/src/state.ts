import { nosync } from "colyseus"
import { EventEmitter } from "eventemitter3"
import { logs } from "./logs"
import { deg2rad, rad2deg, StateEvents } from "@cardsgame/utils"
import { Entity } from "./entity"
import { EntityMap } from "./entityMap"
import { Player } from "./player"

export class State extends EventEmitter {
  // 60 cm
  tableWidth = 600
  // 60 cm
  tableHeight = 600

  entities: Entity

  players = new EntityMap<PlayerData>()
  clients = new EntityMap<string>()
  currentPlayerIdx: number
  get currentPlayer(): PlayerData {
    return this.players[this.currentPlayerIdx]
  }
  get playersCount(): number {
    return this.players.length
  }

  // TODO: think this through:
  gameVariants: any

  isGameStarted = false
  @nosync
  _lastID = -1
  @nosync
  _allEntities = new Map<number, Entity>()

  constructor(options?: IStateOptions) {
    super()
    // TODO: do something with these options.

    this.entities = new Entity({
      state: this,
      type: "root",
      name: "root"
    })

    this.setupListeners()
  }

  setupListeners() {
    this.on(StateEvents.privatePropsSyncRequest, (client: string) => {
      // Bubble it down to every entity
      logs.info("State.privatePropsSyncRequest")
      this.entities.emit(StateEvents.privatePropsSyncRequest, client)
    })
  }

  /**
   * Registers new entity to the gamestate
   * @param entity
   * @returns new ID to be assigned to that entity
   */
  rememberEntity(entity: Entity) {
    const newID = ++this._lastID
    this._allEntities.set(newID, entity)
    return newID
  }

  /**
   * Get an Entity by its ID
   * @param id
   */
  getEntity(id: EntityID): Entity
  /**
   * Get an Entity by its idx path
   * @param path
   */
  getEntity(path: number[]): Entity
  getEntity(idOrPath: EntityID | number[]): Entity {
    if (Array.isArray(idOrPath)) {
      const travel = (entity: Entity, path: number[]) => {
        const idx = path.shift()
        const newChild = entity.children[idx] as Entity
        if (!newChild) {
          logs.error("getEntity/path", `This entity doesn't have such child`)
        }
        if (path.length > 0) {
          return travel(newChild, path)
        } else {
          return newChild
        }
      }
      return travel(this.entities, [...idOrPath])
    }
    return this._allEntities.get(idOrPath)
  }

  /**
   * Gets an array of all entities from the top-most parent
   * to the lowest of the child.
   */
  getEntitiesAlongPath(path: number[]): Entity[] {
    const travel = (entity: Entity, path: number[], result: Entity[] = []) => {
      const idx = path.shift()
      const newChild = entity.children[idx] as Entity
      if (!newChild) {
        logs.error("getEntity/path", `This entity doesn't have such child`)
      }
      result.push(newChild)
      if (path.length > 0) {
        return travel(newChild, path, result)
      } else {
        return result
      }
    }
    return travel(this.entities, [...path])
  }

  logTreeState() {
    logs.log("")
    const indent = (level: number) => {
      return "│ ".repeat(level)
    }
    const travel = (entity: Entity, level: number = 0) => {
      entity.childrenArray.map((child, idx, entities) => {
        if (
          child.parentEntity.isContainer &&
          entities.length > 5 &&
          idx < entities.length - 5
        ) {
          // That's too much, man!
          if (idx === 0) {
            logs.log(`${indent(level)}`, "...")
          }
          return
        }

        const owner = child.owner

        const lastChild = entities.length - 1 === idx
        const sIdx = idx === child.idx ? `${idx}` : `e${child.idx}:s${idx}`
        const sVisibility = child.visibleToPublic ? "Pub" : ""

        const sChildren = child.length > 0 ? child.length + " children" : ""
        const sOwner = owner ? `(${owner.type} ${owner.clientID})` : ""
        const branchSymbol = lastChild ? "┕━" : "┝━"

        logs.log(
          `${indent(level)}${branchSymbol}[${sIdx}]${sVisibility}`,
          `${child.type}:${child.name}-[${child.idx}]`,
          sChildren,
          sOwner
        )
        if (child.length > 0) {
          travel(child, level + 1)
        }
      })
    }
    const root = this.entities
    logs.log(
      `┍━[${root.idx}]`,
      `${root.type}:${root.name}`,
      root.length + "children"
    )
    travel(root, 1)
  }

  static ROOT_ID = 0
}

// Get rid of EventEmitter stuff from the client
nosync(State.prototype, "_events")
nosync(State.prototype, "_eventsCount")
nosync(State.prototype, "_maxListeners")
nosync(State.prototype, "domain")

export interface IStateOptions {
  minClients: number
  maxClients: number
  hostID: string
}

export type PlayerData = {
  entity: Entity
  clientID: string
}
