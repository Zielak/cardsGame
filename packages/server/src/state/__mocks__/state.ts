import type { Player } from "../../player/player.js"

export class State<
  V extends Record<string, unknown> = Record<string, unknown>
> {
  name = "Unnamed"
  type = "state"
  tableWidth = 60 // 60 cm
  tableHeight = 60 // 60 cm
  clients = new Array<string>()
  turnBased = true
  round = 0
  players = new Array<Player>()
  currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }
  isGameStarted = false
  isGameOver = false

  ui = {}

  playerViewPosition = {
    alignX: "center",
    alignY: "bottom",
    paddingX: 0,
    paddingY: 0,
  }

  variantData: V
  hijacksInteractionTarget = false

  id = 0

  _registerEntity(): void {
    // do nothing
  }
}
