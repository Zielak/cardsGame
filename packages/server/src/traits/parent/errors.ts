import type { ChildTrait } from "../child.js"

const INDEX_DOESNT_FIT_RANGE = (desired: number, max: number): string =>
  `idex is not in available range, desired:${desired}, max:${max}`

const acPrefix = "addChild | "
const ADDCHILD_MISSING_ARG0 = (): string =>
  acPrefix + `missing required argument "entity"`
const ADDCHILD_NO_EMPTY_SPOTS = (): string =>
  acPrefix + "no more empty spots to place the item in."
const ADDCHILD_INDEX_OCCUPIED = (): string => acPrefix + "index occupied"
const ADDCHILD_UNEXPECTED_NO_SPOT = (): string =>
  acPrefix + "shouldn't happen! can't actually find any empty spot"

const mtPrefix = "moveChildTo | "
const MOVECHILDTO_NOTHING_TO_MOVE = (idx: number): string =>
  mtPrefix + `I don't have ${idx} child, nothing to move.`
const MOVECHILDTO_MOVE_FAILED = (
  from: number,
  to: number,
  innerError: string
): string => mtPrefix + `${from}->${to} failed: ${innerError}`

const VERY_UNEXPECTED_ERROR = (
  child: ChildTrait,
  removedChild: ChildTrait
): string =>
  `How the f**k did that happen?
  child: ${child.idx}:${child["name"]}
  removedChild: ${removedChild.idx}:${removedChild["name"]}`

export const errors = {
  ADDCHILD_MISSING_ARG0,
  ADDCHILD_NO_EMPTY_SPOTS,
  ADDCHILD_INDEX_OCCUPIED,
  VERY_UNEXPECTED_ERROR,
  ADDCHILD_UNEXPECTED_NO_SPOT,
  INDEX_DOESNT_FIT_RANGE,
  MOVECHILDTO_NOTHING_TO_MOVE,
  MOVECHILDTO_MOVE_FAILED,
}
