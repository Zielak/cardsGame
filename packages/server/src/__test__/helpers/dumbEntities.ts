import { canBeChild } from "@/annotations/canBeChild.js"
import { containsChildren } from "@/annotations/containsChildren.js"
import { ChildTrait } from "@/traits/child.js"
import { applyTraitsMixins, Entity } from "@/traits/entity.js"
import { IdentityTrait } from "@/traits/identity.js"
import { ParentTrait } from "@/traits/parent.js"

@canBeChild
@containsChildren
@applyTraitsMixins([IdentityTrait, ParentTrait, ChildTrait])
export class DumbParent extends Entity<DumbParent> {}
export interface DumbParent extends IdentityTrait, ParentTrait, ChildTrait {}

@canBeChild
@applyTraitsMixins([IdentityTrait, ChildTrait])
export class DumbEntity extends Entity<DumbEntity> {}
export interface DumbEntity extends IdentityTrait, ChildTrait {}
