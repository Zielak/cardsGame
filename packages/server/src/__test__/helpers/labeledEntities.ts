import { canBeChild } from "../../annotations/canBeChild.js"
import { containsChildren } from "../../annotations/containsChildren.js"
import { ChildTrait } from "../../traits/child.js"
import { applyTraitsMixins, Entity } from "../../traits/entity.js"
import { IdentityTrait } from "../../traits/identity.js"
import { LabelTrait } from "../../traits/label.js"
import { ParentTrait } from "../../traits/parent.js"

@canBeChild
@containsChildren
@applyTraitsMixins([IdentityTrait, ParentTrait, ChildTrait, LabelTrait])
export class LabeledParent extends Entity<LabeledParent> {}

export interface LabeledParent
  extends IdentityTrait,
    ParentTrait,
    ChildTrait,
    LabelTrait {}

@canBeChild
@applyTraitsMixins([IdentityTrait, ChildTrait, LabelTrait])
export class LabeledEntity extends Entity<LabeledEntity> {}

export interface LabeledEntity extends IdentityTrait, ChildTrait, LabelTrait {}
