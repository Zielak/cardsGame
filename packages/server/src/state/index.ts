import { ArraySchema, MapSchema } from "@colyseus/schema"

import { containsChildren } from "../annotations/containsChildren"
import { type } from "../annotations/type"
import { Player } from "../players/player"
import { PlayerViewPosition } from "../playerViewPosition"
import { applyTraitsMixins, Entity } from "../traits/entity"
import { IdentityTrait } from "../traits/identity"
import { LabelTrait } from "../traits/label"
import { ParentTrait } from "../traits/parent"

@containsChildren()
@applyTraitsMixins([IdentityTrait, LabelTrait, ParentTrait])
export class State extends Entity<Record<string, unknown>> {
  type = "state"

  @type("number") tableWidth = 60 // 60 cm

  @type("number") tableHeight = 60 // 60 cm

  /**
   * IDs of clients who are connecting to the room.
   * A "client" is someone who just stopped by in this room
   * and not necessarily someone who is playing the game.
   */
  @type(["string"]) clients = new ArraySchema<string>()

  /**
   * Games are turn-based by default. Each player takes their turns one by one.
   * Set this to `false` to allow simultaneous play.
   * Don't rely on `currentPlayer` value for non turn-based games.
   */
  @type("boolean") turnBased = true

  /**
   * Current round number. Increased using `NextRound` command.
   */
  @type("uint16") round = 0

  /**
   * List of players - game participants, after the game starts.
   *
   * TODO: Make players have `idx` if you ever need them to change order mid-game
   */
  @type([Player]) players = new ArraySchema<Player>()

  @type("uint8") currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }

  @type("boolean") isGameStarted = false
  @type("boolean") isGameOver = false

  /**
   * Describes which client-side UI component to reveal to which player.
   * KEY = uiElement, VALUE = clientID
   *
   * NOTE: It's a very simple example, for more advanced use (eg. UI element
   * visible for multiple certain players) feel free to
   * create your own ui handler in your State
   */
  @type({ map: "string" }) ui = new MapSchema<string>()

  /**
   * A construct describing how should player's "focused" items
   * be positioned in their view. Containers of other players
   * will not follow these rules.
   * Default is: center/bottom.
   */
  @type(PlayerViewPosition) playerViewPosition = new PlayerViewPosition()

  /**
   * ID of last known, registered Entity.
   */
  private _lastID = -1

  /**
   * Map of every Entity in this state.
   */
  private readonly _allEntities = new Map<number, IdentityTrait>()

  constructor() {
    super(undefined)

    this.hijacksInteractionTarget = false
  }

  /**
   * Registers new entity to the game state.
   * @param entity
   * @private internal use only
   */
  _registerEntity(entity: IdentityTrait): void {
    if (this._allEntities.has(entity.id)) {
      if (this._allEntities.get(entity.id) !== entity) {
        throw new Error(
          "State: registering entity, which was already registered WITH DIFFERENT ID?"
        )
      }
      return
    }

    this._lastID++
    const newID = this._lastID

    Object.defineProperty(entity, "id", {
      configurable: false,
      writable: false,
      value: newID,
    })
    this._allEntities.set(newID, entity)
  }
}

interface Mixin extends IdentityTrait, LabelTrait, ParentTrait {}

export interface State extends Mixin {}
