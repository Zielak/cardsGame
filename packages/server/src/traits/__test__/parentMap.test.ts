import { DumbEntity, DumbParent } from "../../__test__/helpers/dumbEntities.js"
import { LabeledEntity } from "../../__test__/helpers/labeledEntities.js"
import { State } from "../../state/state.js"
import type { ChildTrait } from "../child.js"
import type { IdentityTrait } from "../identity.js"
import { errors } from "../parent/errors.js"

let state: State
const OPTIONS: Partial<DumbParent> = { collectionBehaviour: "map" }

beforeEach(() => {
  state = new State()
})

describe("ParentConstructor", () => {
  test("sets default properties", () => {
    const parent = new DumbParent(state, OPTIONS)

    expect(parent.countChildren()).toBe(0)
    expect(parent.hijacksInteractionTarget).toBe(false)
    expect(parent.maxChildren).toBe(Infinity)
  })
  test("remembers custom values", () => {
    const parent = new DumbParent(state, { ...OPTIONS, maxChildren: 3 })

    expect(parent.maxChildren).toBe(3)
  })
})

describe("#removeChildAt", () => {
  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state, OPTIONS)
  })
  it(`throws on invalid "idx"`, () => {
    expect(parent.removeChildAt(-2)).toBe(false)
    expect(parent.removeChildAt(NaN)).toBe(false)
    expect(parent.removeChildAt(Infinity)).toBe(false)
    expect(parent.removeChildAt(10)).toBe(false)
  })
  it("removes the only child", () => {
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChildAt(0)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })
  it("removes child in the middle", () => {
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(3)
    expect(parent.removeChildAt(1)).toBe(true)
    expect(parent.countChildren()).toBe(2)
  })
})

describe("#removeChild", () => {
  let parent: DumbParent
  it("calls removeChildAt", () => {
    parent = new DumbParent(state, OPTIONS)
    const entity = new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChild(entity)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })

  it("finds proper child", () => {
    parent = new DumbParent(state, OPTIONS)
    new DumbEntity(state, { parent })
    const entity = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    parent.removeChild(entity)
    expect(
      parent.getChildren().map((e: ChildTrait & IdentityTrait) => e.id)
    ).toMatchSnapshot()
  })

  it("returns false for not its child", () => {
    const child = new DumbEntity(state)

    expect(parent.removeChild(child)).toBe(false)
  })
})

describe("#addChild", () => {
  let parent: DumbParent
  let entity: DumbEntity
  beforeEach(() => {
    parent = new DumbParent(state, { ...OPTIONS, maxChildren: 5 })
    new DumbEntity(state, { parent })
  })
  it("adds new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(1)
    expect(parent.countChildren()).toBe(2)
  })
  it("prepends new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity, true)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(0)
    expect(parent.countChildren()).toBe(2)
  })
  it("prepends in busier parent", () => {
    // Clean parent this time.
    parent = new DumbParent(state, OPTIONS)
    const e0 = new LabeledEntity(state, { parent, name: "e0", idx: 0 })
    const e1 = new LabeledEntity(state, { parent, name: "e1", idx: 1 })
    const e3 = new LabeledEntity(state, { parent, name: "e3", idx: 3 })
    entity = new DumbEntity(state)

    expect(parent.getChildren().map((c) => c.idx)).toStrictEqual([0, 1, 3])
    expect(e0.idx).toBe(0)
    expect(e1.idx).toBe(1)
    expect(e3.idx).toBe(3)

    parent.addChild(entity, true)

    expect(entity.parent).toBe(parent)
    expect(parent.getChildren().map((c) => c.idx)).toStrictEqual([0, 1, 2, 3])

    expect(entity.idx).toBe(0)
    expect(e0.idx).toBe(1)
    expect(e1.idx).toBe(2)
    expect(e3.idx).toBe(3)

    expect(parent.countChildren()).toBe(4)
  })
  it("throws at out of bounds index", () => {
    expect(() => parent.addChild(new DumbEntity(state), 10)).toThrow()
  })
  it("add item at index", () => {
    const entityA = new DumbEntity(state)
    const entityB = new DumbEntity(state)

    // index 0 is already set at beforeEach()
    expect(() => parent.addChild(entityA, 0)).toThrow()

    parent.addChild(entityA, 3)

    expect(entityA.parent).toBe(parent)
    expect(entityA.idx).toBe(3)
    expect(parent.countChildren()).toBe(2)

    parent.addChild(entityB, 4)

    expect(entityB.parent).toBe(parent)
    expect(entityB.idx).toBe(4)
    expect(parent.countChildren()).toBe(3)
  })
  it("throws at full parent", () => {
    parent = new DumbParent(state, { ...OPTIONS, maxChildren: 1 })

    new DumbEntity(state, { parent })

    expect(() => parent.addChild(new DumbEntity(state))).toThrow(
      errors.ADDCHILD_NO_EMPTY_SPOTS()
    )
  })
})

describe("#moveChildTo", () => {
  const mapChildren = (parent: DumbParent): any[] => {
    const max =
      parent.maxChildren !== Infinity
        ? parent.maxChildren
        : parent.countChildren()

    const results = []
    for (let i = 0; i < max; i++) {
      const child = parent.getChild<DumbEntity>(i)
      results.push(child ? child.id : null)
    }
    return results
  }

  let parent: DumbParent

  describe("filled", () => {
    beforeEach(() => {
      parent = new DumbParent(state, { ...OPTIONS, maxChildren: 5 })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })
    })
    it("moves item3 down 1 position", () => {
      parent.moveChildTo(2, 1)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 up 1 position", () => {
      parent.moveChildTo(2, 3)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 to bottom", () => {
      parent.moveChildTo(2, 0)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item3 to top", () => {
      parent.moveChildTo(2, parent.countChildren() - 1)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("doesn't move anything (item3 to same position)", () => {
      parent.moveChildTo(2, 2)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("throws on indexes out of boundaries", () => {
      expect(() => parent.moveChildTo(2, 10)).toThrow()
      expect(() => parent.moveChildTo(10, 2)).toThrow()
      expect(() => parent.moveChildTo(-2, 0)).toThrow()
      expect(() => parent.moveChildTo(0, -1)).toThrow()

      expect(mapChildren(parent)).toMatchSnapshot()
    })
  })

  describe("with empty spaces", () => {
    beforeEach(() => {
      parent = new DumbParent(state, { ...OPTIONS, maxChildren: 5 })
      new DumbEntity(state, { parent, idx: 2 })
    })
    it("moves item down 1 position", () => {
      parent.moveChildTo(2, 1)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item up 1 position", () => {
      parent.moveChildTo(2, 3)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item to bottom", () => {
      parent.moveChildTo(2, 0)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("moves item to top", () => {
      parent.moveChildTo(2, parent.maxChildren - 1)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("doesn't move anything (item to same position)", () => {
      parent.moveChildTo(2, 2)
      expect(mapChildren(parent)).toMatchSnapshot()
    })
    it("throws on indexes out of boundaries", () => {
      expect(() => parent.moveChildTo(2, 10)).toThrow()
      expect(() => parent.moveChildTo(10, 2)).toThrow()
      expect(() => parent.moveChildTo(-2, 0)).toThrow()
      expect(() => parent.moveChildTo(0, -2)).toThrow()

      expect(mapChildren(parent)).toMatchSnapshot()
    })
  })
})

test.todo("#countChildren")
describe("#getChildren", () => {
  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state, OPTIONS)
  })
  test("empty parent", () => {
    expect(parent.getChildren()).toMatchSnapshot()
  })
  test("couple of children", () => {
    const first = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    const last = new DumbEntity(state, { parent })

    const result = parent.getChildren()
    expect(result).toMatchSnapshot()
    expect(result[0]).toBe(first)
    expect(result[result.length - 1]).toBe(last)
  })
})

describe("#getChild", () => {
  let parent
  let entity

  it("gets child", () => {
    parent = new DumbParent(state, OPTIONS)

    new DumbEntity(state, { parent })
    entity = new DumbEntity(state, { parent })

    expect(parent.getChild(1)).toBe(entity)
  })
  test("idx out of range", () => {
    parent = new DumbParent(state, OPTIONS)

    expect(parent.getChild(-1)).toBeUndefined()
    expect(parent.getChild(0)).toBeUndefined()
    expect(parent.getChild(Infinity)).toBeUndefined()
  })
  test("idx out of range in limited container", () => {
    parent = new DumbParent(state, { ...OPTIONS, maxChildren: 5 })

    expect(parent.getChild(-1)).toBeUndefined()
    expect(parent.getChild(10)).toBeUndefined()
    expect(parent.getChild(-Infinity)).toBeUndefined()
    expect(parent.getChild(Infinity)).toBeUndefined()
  })
})

test("#getTop", () => {
  const parent = new DumbParent(state, OPTIONS)
  new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })
  const entity = new DumbEntity(state, { parent })

  expect(parent.getTop()).toBe(entity)
})

test("#getBottom", () => {
  const parent = new DumbParent(state, OPTIONS)
  const entity = new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })

  expect(parent.getBottom()).toBe(entity)
})

describe("#getFirstEmptySpot", () => {
  let parent: DumbParent
  describe("limited parent", () => {
    beforeEach(() => {
      parent = new DumbParent(state, { ...OPTIONS, maxChildren: 3 })
    })
    it("returns zero on empty parent", () => {
      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return zero", () => {
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return one", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(1)
    })
    it("return last", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })

      expect(parent.getFirstEmptySpot()).toBe(2)
    })
    it("can't find it so it returns -1", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(-1)
    })
  })
  describe("unlimited parent", () => {
    beforeEach(() => {
      parent = new DumbParent(state, OPTIONS)
    })
    it("returns zero on empty parent", () => {
      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return zero", () => {
      new DumbEntity(state, { parent, idx: 1 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(0)
    })
    it("return one", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 2 })

      expect(parent.getFirstEmptySpot()).toBe(1)
    })
    it("return last", () => {
      new DumbEntity(state, { parent, idx: 0 })
      new DumbEntity(state, { parent, idx: 1 })

      expect(parent.getFirstEmptySpot()).toBe(2)
    })
  })
})

describe("#getLastEmptySpot", () => {
  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state, { ...OPTIONS, maxChildren: 5 })
  })
  it("picks last on empty parent", () => {
    expect(parent.getLastEmptySpot()).toBe(4)
  })
  it("picks last spot", () => {
    // 0
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    // 3
    // 4

    expect(parent.getLastEmptySpot()).toBe(4)
  })
  it("picks 3 from 2 empty spaces", () => {
    // 0
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    // 3
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(3)
  })
  it("picks 2 from 2 empty spaces", () => {
    new DumbEntity(state, { parent, idx: 0 })
    // 1
    // 2
    new DumbEntity(state, { parent, idx: 3 })
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(2)
  })
  it("returns -1, for filled container", () => {
    new DumbEntity(state, { parent, idx: 0 })
    new DumbEntity(state, { parent, idx: 1 })
    new DumbEntity(state, { parent, idx: 2 })
    new DumbEntity(state, { parent, idx: 3 })
    new DumbEntity(state, { parent, idx: 4 })

    expect(parent.getLastEmptySpot()).toBe(-1)
  })
})
