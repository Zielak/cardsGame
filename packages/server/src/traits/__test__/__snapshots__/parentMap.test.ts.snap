// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getChildren couple of children 1`] = `
[
  {
    "idx": 0,
  },
  {
    "idx": 1,
  },
  {
    "idx": 2,
  },
  {
    "idx": 3,
  },
  {
    "idx": 4,
  },
]
`;

exports[`#getChildren empty parent 1`] = `[]`;

exports[`#moveChildTo filled doesn't move anything (item3 to same position) 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 down 1 position 1`] = `
[
  1,
  3,
  2,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 to bottom 1`] = `
[
  3,
  2,
  1,
  4,
  5,
]
`;

exports[`#moveChildTo filled moves item3 to top 1`] = `
[
  1,
  2,
  5,
  4,
  3,
]
`;

exports[`#moveChildTo filled moves item3 up 1 position 1`] = `
[
  1,
  2,
  4,
  3,
  5,
]
`;

exports[`#moveChildTo filled throws on indexes out of boundaries 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`#moveChildTo with empty spaces doesn't move anything (item to same position) 1`] = `
[
  null,
  null,
  1,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item down 1 position 1`] = `
[
  null,
  1,
  null,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item to bottom 1`] = `
[
  1,
  null,
  null,
  null,
  null,
]
`;

exports[`#moveChildTo with empty spaces moves item to top 1`] = `
[
  null,
  null,
  null,
  null,
  1,
]
`;

exports[`#moveChildTo with empty spaces moves item up 1 position 1`] = `
[
  null,
  null,
  null,
  1,
  null,
]
`;

exports[`#moveChildTo with empty spaces throws on indexes out of boundaries 1`] = `
[
  null,
  null,
  1,
  null,
  null,
]
`;

exports[`#removeChild finds proper child 1`] = `
[
  1,
  3,
]
`;
